File: Golang/92.反转链表 II.go
package leetcode

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseBetween(head *ListNode, left int, right int) *ListNode {
	var dummy *ListNode = &ListNode{Val: 0, Next: head}
	var pre *ListNode = dummy
	for i := 0; i < left-1; i++ {
		pre = pre.Next
	}
	var curr *ListNode = pre.Next
	var lat *ListNode = curr.Next
	for i := left; i < right; i++ {
		curr.Next = lat.Next
		lat.Next = pre.Next
		pre.Next = lat
		lat = curr.Next
	}
	return dummy.Next
}

File: TypeScript/92.反转链表 II.ts
// Definition for singly-linked list.
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}


function reverseBetween(head: ListNode | null, left: number, right: number): ListNode | null {
    const dummy = new ListNode(0, head)
    let pre = dummy
    for (let i = 0; i < left - 1; ++i) {
        pre = pre.next
    }
    let curr = pre.next
    let lat = curr.next
    for (let i = left; i < right; ++i) {
        curr.next = lat.next
        lat.next = pre.next
        pre.next = lat
        lat = curr.next
    }
    return dummy.next
};
